File: samsung/BOJ17779.py
from sys import stdin
from collections import deque
readline = stdin.readline

ans = int(1e6)
N = int(readline().strip())
_map = [[0] * (N+1)]
total = 0
for _ in range(N):
    line = [0]+list(map(int, readline().split()))
    total += sum(line)
    _map.append(line)


def solve():
    for x in range(1, N+1):
        for y in range(1, N+1):
            for d1 in range(1, N+1):
                for d2 in range(1, N+1):
                    if x + d1 + d2 > N:
                        continue
                    if y - d1 < 1:
                        continue
                    if y + d2 > N:
                        continue
                    find(x, y, d1, d2)                
def find(x, y, d1, d2):
    global ans
    region = [[0] * (N+1) for _ in range(N+1)]    
    population = [0] * 6

    # 경계선
    region[x][y] = 5
    for i in range(1, d1+1):
        region[x+i][y-i] = 5
    
    for i in range(1, d2+1):
        region[x+i][y+i] = 5

    for i in range(1, d2+1):
        region[x+d1+i][y-d1+i] = 5

    for i in range(1, d1+1):
        region[x+d2+i][y+d2-i] = 5

    # 1 구역
    for r in range(1, x+d1):
        for c in range(1, y+1):
            if region[r][c] == 5:break
            if region[r][c] == 0:
                region[r][c] = 1
                population[1] += _map[r][c]
    # 2 구역
    for r in range(1, x+d2+1):
        for c in range(N, y, -1):
            if region[r][c] == 5:break
            if region[r][c] == 0:
                region[r][c] = 2
                population[2] += _map[r][c]

    # 3 구역
    for r in range(x+d1, N+1):
        for c in range(1, y-d1+d2):
            if region[r][c] == 5:break
            if region[r][c] == 0:
                region[r][c] = 3
                population[3] += _map[r][c]

    # 4 구역
    for r in range(x+d2+1, N+1):
        for c in range(N, y-d1+d2-1, -1):
            if region[r][c] == 5:break
            if region[r][c] == 0:
                region[r][c] = 4
                population[4] += _map[r][c]
    
    population[5] = total - sum(population)

    ans = min(ans, max(population[1:])-min(population[1:]))

solve()
print(ans)

File: samsung/java/BOJ17779.java
import java.util.*;
import java.io.*;

public class BOJ17779{
    static int N;
    static int[][] map;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        N = Integer.parseInt(br.readLine());
        map = new int[N+1][N+1];
        int ans = 100000000;
        int total = 0;

        for(int i=1;i<=N;i++){
            String[] line = br.readLine().split(" ");
            for(int j=1;j<=N;j++){
                map[i][j] = Integer.parseInt(line[j-1]);
                total += map[i][j];
            }
        }
        
        for(int x=1;x<=N;x++){
            for(int y=1;y<=N;y++){
                for(int d1=1;d1<=N;d1++){
                    for(int d2=1;d2<=N;d2++){
                        if (x + d1 + d2 > N) continue;
                        if (y - d1 < 1) continue;
                        if (y + d2 > N) continue;

                        ans = Math.min(ans, find(x, y, d1, d2, total));
                    }
                }
            }
        }
        System.out.println(ans);
    }

    public static int find(int x, int y, int d1, int d2, int total){
        int[][] region = new int[N+1][N+1];
        int[] population = new int[N+1];
        
        // 경계
        region[x][y] = 5;
        for(int i=1;i<=d1;i++){
            region[x+i][y-i] = 5; 
        }        
        for(int i=1;i<=d2;i++){
            region[x+i][y+i] = 5; 
        }        
        for(int i=1;i<=d2;i++){
            region[x+d1+i][y-d1+i] = 5;
        }
        for(int i=1;i<=d1;i++){
            region[x+d2+i][y+d2-i] = 5;
        }

        // 1 구역
        for(int r=1;r<x+d1;r++){
            for(int c=1;c<=y;c++){
                if(region[r][c] == 5) break;
                if(region[r][c] == 0){
                    region[r][c] = 1;
                    population[1] += map[r][c];
            
                }
            }
        }
        
        // 2 구역
        for(int r=1;r<=x+d2;r++){
            for(int c=N;c>y;c--){
                if(region[r][c] == 5) break;
                if(region[r][c] == 0){
                    region[r][c] = 2;
                    population[2] += map[r][c];
                }
            }
        }
        
        // 3 구역
        for(int r=x+d1;r<=N;r++){
            for(int c=1;c<y-d1+d2;c++){
                if(region[r][c] == 5) break;
                if(region[r][c] == 0){
                    region[r][c] = 3;
                    population[3] += map[r][c];
                }
            }
        }

        // 4 구역
        for(int r=x+d2+1;r<=N;r++){
            for(int c=N;c>=y-d1+d2;c--){
                if(region[r][c] == 5) break;
                if(region[r][c] == 0){
                    region[r][c] = 4;
                    population[4] += map[r][c];
                }
            }
        }
        
        int cur = 0;
        for(int p : population){
            cur += p;
        }
        population[5] = total-cur; 

        int max = 0;
        int min = 100000000;
        for(int i=1;i<=5;i++){
            if(max < population[i]){
                max = population[i];
            }
            if(min > population[i]){
                min = population[i];
            }
        }
        return max - min;
    }
}